alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/processor/notification/NotificationProcessor.java

```diff
@@ -71,18 +76,42 @@ public NotificationProcessor(ConfigurationAccessor configurationAccessor, JobAcc
 
     public List<DistributionEvent> processNotifications(FrequencyType frequency, List<AlertNotificationModel> notifications) {
         logger.info("Notifications to Process: {}", notifications.size());
-        List<ConfigurationJobModel> jobsForFrequency = jobAccessor.getJobsByFrequency(frequency);
-        return processNotificationsForJobs(jobsForFrequency, notifications);
+        SetMap<NotificationFilterModel, AlertNotificationModel> notificationFilterMap = extractNotificationInformation(notifications);
+
+        List<DistributionEvent> events = new ArrayList<>();
+        for (Map.Entry<NotificationFilterModel, Set<AlertNotificationModel>> entry : notificationFilterMap.entrySet()) {
+            NotificationFilterModel notificationFilterModel = entry.getKey();
+            List<ConfigurationJobModel> matchingJobs = jobAccessor.getMatchingEnabledJobs(frequency, notificationFilterModel.getProviderConfigId(), notificationFilterModel.getNotificationType());
+
+            List<AlertNotificationModel> matchingNotifications = new ArrayList<>(entry.getValue());
+            if (!matchingNotifications.isEmpty() && !matchingJobs.isEmpty()) {
+                events.addAll(processNotificationsThatMatchFilter(notificationFilterModel, matchingJobs, matchingNotifications));
+            }
+        }
+        return events;
     }
 
     public List<DistributionEvent> processNotifications(List<AlertNotificationModel> notifications) {
+        // used in AuditEntryActions
         // when a job is deleted use this method to send the same notification to the current set of jobs. i.e. audit
         // FIXME consider paging
-        List<ConfigurationJobModel> allJobs = jobAccessor.getAllJobs();
-        return processNotificationsForJobs(allJobs, notifications);
+        SetMap<NotificationFilterModel, AlertNotificationModel> notificationFilterMap = extractNotificationInformation(notifications);
+
+        List<DistributionEvent> events = new ArrayList<>();
+        for (Map.Entry<NotificationFilterModel, Set<AlertNotificationModel>> entry : notificationFilterMap.entrySet()) {
+            NotificationFilterModel notificationFilterModel = entry.getKey();
+            List<ConfigurationJobModel> matchingJobs = jobAccessor.getMatchingEnabledJobs(notificationFilterModel.getProviderConfigId(), notificationFilterModel.getNotificationType());
+
+            List<AlertNotificationModel> matchingNotifications = new ArrayList<>(entry.getValue());
+            if (!matchingNotifications.isEmpty() && !matchingJobs.isEmpty()) {
+                events.addAll(processNotificationsThatMatchFilter(notificationFilterModel, matchingJobs, matchingNotifications));
+            }
+        }
+        return events;
```

Comment:
this method is pretty much a duplicate of 79. I think you can create a private method that takes in a bifunction that performs the getMatchingEnabledJobs functionality.

Reply:
I will add an if statement to the method that takes the FrequencyType parameter and just have the other method call that. I can't use a Bifunction as you described because the `getMatchingEnabledJobs` in each method have a different number of parameters.